tests/library/qemu-user/conftest.py

```diff
@@ -59,6 +58,65 @@
 }
 
 
+def reserve_port(ip: str = "127.0.0.1", port: int = 0) -> str:
+    """
+    https://github.com/Yelp/ephemeral-port-reserve/blob/master/ephemeral_port_reserve.py
+
+    Bind to an ephemeral port, force it into the TIME_WAIT state, and unbind it.
+
+    This means that further ephemeral port alloctions won't pick this "reserved" port,
+    but subprocesses can still bind to it explicitly, given that they use SO_REUSEADDR.
+    By default on linux you have a grace period of 60 seconds to reuse this port.
+    To check your own particular value:
+    $ cat /proc/sys/net/ipv4/tcp_fin_timeout
+    60
+
+    By default, the port will be reserved for localhost (aka 127.0.0.1).
+    To reserve a port for a different ip, provide the ip as the first argument.
+    Note that IP 0.0.0.0 is interpreted as localhost.
+    """
+    import contextlib
+    import errno
+    from socket import SO_REUSEADDR
+    from socket import SOL_SOCKET
+    from socket import error as SocketError
+    from socket import socket
+
+    port = int(port)
+    with contextlib.closing(socket()) as s:
+        s.setsockopt(SOL_SOCKET, SO_REUSEADDR, 1)
+        try:
+            s.bind((ip, port))
+        except SocketError as e:
+            # socket.error: EADDRINUSE Address already in use
+            if e.errno == errno.EADDRINUSE and port != 0:
+                s.bind((ip, 0))
+            else:
+                raise
+
+        # the connect below deadlocks on kernel >= 4.4.0 unless this arg is greater than zero
+        s.listen(1)
+
+        sockname = s.getsockname()
+
+        # these three are necessary just to get the port into a TIME_WAIT state
+        with contextlib.closing(socket()) as s2:
+            s2.connect(sockname)
+            sock, _ = s.accept()
+            with contextlib.closing(sock):
+                return sockname[1]
+
+
+def ensure_qemu_port():
```

Comment:
why not just pass `reserve_port()` into qemu_assembly_run/qemu_start_binary?
QEMU_PORT can be removed from globals, not used outside this fine

Reply:
> why not just pass reserve_port() into qemu_assembly_run/qemu_start_binary?

Doing it this way took the least amount of changes, and I'm trying to minimize those given how large the PR already is, at least to try to make it less of a chore to review and to ensure it won't break our tests after it's merged.

> QEMU_PORT can be removed from globals, not used outside this file

Yeah, you're right. I'll take it out